### PR TITLE
Add filter to  Item and Collection `get_assets` method

### PR DESCRIPTION
**Pull request by pjhartzell**
**Related Issue(s):** 
#873 

**Description:**
Adds ability to filter Item and Collection Assets by media_type and role via `get_assets()` method.

**PR Checklist:**

- [x] Code is formatted (run `pre-commit run --all-files`)
- [x] Tests pass (run `scripts/test`)
- [x] Documentation has been updated to reflect changes, if applicable
- [x] This PR maintains or improves overall codebase code coverage.
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/main/CHANGELOG.md). See [the docs](https://pystac.readthedocs.io/en/latest/contributing.html#changelog) for information about adding to the changelog.

### CODE REVIEW

File: pystac/collection.py [hunk 1]
```diff
@@ -675,13 +675,37 @@ def from_dict(
...
+                for key, asset in assets.items()
+                if asset.media_type == media_type
+            }
+        if role is not None:
+            assets = {
+                key: asset
+                for key, asset in assets.items()
+                if asset.roles is not None and role in asset.roles
+            }
+        return assets
```
Comment (gadomski) [inline pystac/collection.py:708]: This is a little nitpicky, but we're iterating over the assets as many as three times; once to create the new dict, then twice if both `media_type` and `role` are set. Iterating once seems preferable (with a short circuit):

```python
if media_type is None and role is None:
    return dict(self.assets.items())
assets = dict()
for key, asset in self.assets.items():
    if (media_type is None or asset.media_type == media_type) and (role is None or role in asset.roles):
        assets[key] = asset
return assets
```

File: pystac/item.py [hunk 2]
```diff
@@ -221,13 +221,37 @@ def set_datetime(self, datetime: Datetime, asset: Optional[Asset] = None) -> Non
...
+                for key, asset in assets.items()
+                if asset.media_type == media_type
+            }
+        if role is not None:
+            assets = {
+                key: asset
+                for key, asset in assets.items()
+                if asset.roles is not None and role in asset.roles
+            }
+        return assets
```
Comment (gadomski) [inline pystac/item.py:254]: Same as the comment on the collection method.

**Review by gadomski — CHANGES_REQUESTED** [review]

File: pystac/collection.py (hunk 1 above)
Comment (pjhartzell) [reply pystac/collection.py]: Nope, not nitpicky. That's cleaner. We do need to augment the condition to guard against a `None` value in `asset.roles`:

```python
if (media_type is None or asset.media_type == media_type) and (role is None or (asset.roles is not None and role in asset.roles)):
```

That's getting to be a mouthful. We could simplify `asset.roles is not None and ...` to just `asset.roles and ...`, but that's not as clear (to me).

File: pystac/collection.py [hunk 3]
```diff
@@ -675,13 +675,32 @@ def from_dict(
...
+            Dict[str, Asset]: A dictionary of assets that match ``media_type``
+                and/or ``role`` if set or else all of this collection's assets.
         """
-        return dict(self.assets.items())
+        if media_type is None and role is None:
+            return dict(self.assets.items())
+        assets = dict()
+        for key, asset in self.assets.items():
+            if (media_type is None or asset.media_type == media_type) and (
+                role is None or (asset.roles is not None and role in asset.roles)
```
Comment (gadomski) [inline pystac/collection.py:700]: I agree that this is getting a little wild. What do you think about a new `Asset.has_role` method? E.g.

```suggestion
            if (media_type is None or asset.media_type == media_type) and (role is None or asset.has_role(role)):
```

and then at https://github.com/stac-utils/pystac/blob/017ddfd8592e822ecc6905b5ea3ed0cd121798dc/pystac/asset.py:

```python
def has_role(self, role: str) -> bool:
    if self.roles is None:
        return False
    else:
        return role in self.roles

Comment (pjhartzell) [reply pystac/collection.py]: Yeah, that works for me. Done.

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 ### Added
 
+- Add `media_type` and `role` filtering to Item and Collection `get_assets()` method ([#936](https://github.com/stac-utils/pystac/pull/936))
```
Comment (gadomski) [inline CHANGELOG.md:7]:
```suggestion
- Add `media_type` and `role` filtering to Item and Collection `get_assets()` method ([#936](https://github.com/stac-utils/pystac/pull/936))
- `Asset.has_role` ([#936](https://github.com/stac-utils/pystac/pull/936))
```

**Review by gadomski — CHANGES_REQUESTED** [review]

**Review by gadomski — APPROVED** [review]

File: pystac/collection.py [hunk 5]
```diff
@@ -675,13 +675,32 @@ def from_dict(
 
         return collection
 
-    def get_assets(self) -> Dict[str, Asset]:
-        """Get this item's assets.
+    def get_assets(
+        self,
+        media_type: Optional[Union[str, pystac.MediaType]] = None,
+        role: Optional[Union[str, pystac.MediaType]] = None,
```
Comment (gadomski) [inline pystac/collection.py:681]:
```suggestion
        role: Optional[str] = None,
```

File: pystac/item.py [hunk 6]
```diff
@@ -221,13 +221,32 @@ def set_datetime(self, datetime: Datetime, asset: Optional[Asset] = None) -> Non
         else:
             asset.extra_fields["datetime"] = datetime_to_str(datetime)
 
-    def get_assets(self) -> Dict[str, Asset]:
+    def get_assets(
+        self,
+        media_type: Optional[Union[str, pystac.MediaType]] = None,
+        role: Optional[Union[str, pystac.MediaType]] = None,
```
Comment (gadomski) [inline pystac/item.py:227]:
```suggestion
        role: Optional[str] = None,
```

**Review by gadomski — CHANGES_REQUESTED** [review]

File: pystac/collection.py (hunk 5 above)
Comment (pjhartzell) [reply pystac/collection.py]: Yeesh. Good catch. Fixed.

File: pystac/item.py (hunk 6 above)
Comment (pjhartzell) [reply pystac/item.py]: Fixed.

**Review by gadomski — APPROVED** [review]

**Pull request merged**